Adgr email subscription

anathomical opened the pull request:
## WHAT
- Add front-end to manage email subscriptions for the Admin Diagnostic Report
- Add workers to manage sending Admin Diagnostic Report subscriptions on schedule
## WHY
We want admins to be able to subscribe to regular copies of their Admin Diagnostic Reports, mirroring behavior in the Usage Snapshot report
## HOW
- Add a new mailer for ADGR Subscription Emails, including a new view
- Add new Cron entries to send Weekly and Monthly subscriptions
- Add new types to `AdminReportFilterSelections` for the new subscriptions
- Refactor workers that send on-demand emails so that it can be sub-classed for subscription emails
- Slight refactor the "Subscribe" modal used in the Usage Snapshots report to make it useable in ADGR
- Update ADGR front-end to have a "Subscribe" button and functionality to save or delete subscriptions
- Fix the unsubscribe functionality so that it works for non-logged-in-users from email links

### Screenshots
![image](https://github.com/user-attachments/assets/dfb89a03-e1e5-489f-bfaa-52ccc564b9c3)
![image](https://github.com/user-attachments/assets/cbd69808-79df-4057-98a7-e42a32bb63a8)

### Notion Card Links
https://www.notion.so/quill/1-Frontend-Email-subscription-for-Admin-Diagnostic-Growth-Report-d7d578f870104d77ac9e95b23ee0ab51
https://www.notion.so/quill/1-Backend-scheduled-worker-to-send-Admin-Diagnostic-Growth-Report-email-subscriptions-f26822ef6a324360943b96a9375313fa?pvs=4

### What have you done to QA this feature?
- Deploy to staging
- Click on Subscribe and select "Yes" and "Weekly" and click "Save"
- - Confirm that a new `EmailSubscription` is created in the database for the user with `frequency` == `weekly`
- Click on Subscribe
- - Confirm that "Yes" and "Weekly" are already selected
- Refresh page and repeat previous two steps to confirm database persistence
- Click on Subscribe and Select "Yes" and "Monthly" and click "Save"
- - Confirm that the `EmailSubscription` in the database for the user has `frequency` == `monthly`
- Click on Subscribe and Select "No" and click "Save"
- - Confirm that the `EmailSubscription` in the database for the user has been destroyed
- Re-save a "Weekly" subscription
- Using the Rails console, find an admin account with no current Premium and create a "Weekly" subscription for them
- Use Rails console to trigger the sending of all Weekly subscriptions
- Confirm that the non-Premium admin is not sent an email
- Confirm that email is delivered to the Premium admin that we're testing with
- - Confirm that subject line and message body reference "Weekly" frequency
- - Confirm that body greeting references admin's name
- - Confirm that links to all three CSVs download files
- - - Confirm that each file is the correct one (data from the named tab)
- - - Confirm that each file has data matching what the web report shows
- - Confirm that "Explore more reports" link takes user to the admin hub
- - Confirm that "Unsubscribe" link sends user to an Unsubscribe confirmation page
- - - Confirm that clicking on "Unsubscribe" from the Unsubscribe page destroys the `EmailSubscription` for the user
- - - - Confirm that this works even if the user is logged out

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? | Yes
Have you deployed to Staging? | Yes
Self-Review: Have you done an initial self-review of the code below on Github? | Yes

anathomical committed:
WIP set up new workers to send email subscriptions

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into adgr-email-subscription-worker

anathomical committed:
Add a worker for Monthly subscriptions and add workers to Cron

anathomical committed:
Refactor workers and add specs

anathomical committed:
Do some refactoring to make mailers construct more cleanly

anathomical committed:
Update cron spec

anathomical committed:
WIP adding subscription front-end to ADGR

anathomical committed:
Merge branch 'adgr-email-subscription-worker' into adgr-subscription-frontend

anathomical committed:
WIP integrating backend for subscriptions

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into adgr-subscription-frontend

anathomical committed:
Finish front-end for ADGR subscriptions

anathomical committed:
Fix super call to pass appropriate params

anathomical committed:
Update default Diagnostic ID when not selected

anathomical committed:
Fix mis-configuration in unsubscribe controller

anathomical committed:
Update unsubscribe link to work

anathomical committed:
Set frequency for email template

anathomical committed:
Add some debug output to chase down CSV generation failure

anathomical committed:
Remove debugging output

anathomical committed:
Remove unused constant declarations

anathomical committed:
Fix linting errors

anathomical committed:
Update snapshot

anathomical committed:
More snapshot cleanup

anathomical committed:
Update constants in spec since we've standardized them

anathomical committed:
Update tests with appropriate new param names

anathomical committed:
Update worker and spec to handle premium requirement

anathomical commented on services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/index.tsx:
```diff
@@ -177,11 +197,59 @@ export const DiagnosticGrowthReportsContainer = ({
+    setIsSubscriptionModalOpen(false);
+  }
+
+  function handleSubscriptionSave(isSubscribed, frequency) {
+    setIsSubscriptionModalOpen(false);
+    if (isSubscribed) {
+      requestPost('/admin_report_filter_selections/show', { report: `${FILTER_SELECTIONS_REPORT_BASE.slice(0,-1)}` }, (sharedSelections) => {
+        requestPost('/admin_report_filter_selections/show', { report: `${FILTER_SELECTIONS_REPORT_BASE}${OVERVIEW}` }, (overviewSelections) => {
+          requestPost('/admin_report_filter_selections/show', { report: `${FILTER_SELECTIONS_REPORT_BASE}${SKILL}` }, (skillSelections) => {
+            requestPost('/admin_report_filter_selections/show', { report: `${FILTER_SELECTIONS_REPORT_BASE}${STUDENT}` }, (studentSelections) => {
+              if (sharedSelections) postFilterSelections(`subscription`, sharedSelections.filter_selections)
+              if (overviewSelections) postFilterSelections(`subscription_${OVERVIEW}`, overviewSelections.filter_selections)
+              if (skillSelections) postFilterSelections(`subscription_${SKILL}`, skillSelections.filter_selections)
+              if (studentSelections) postFilterSelections(`subscription_${STUDENT}`, studentSelections.filter_selections)
+
+              createOrUpdateEmailSubscription(frequency)
+            })
+          })
+        })
+      })
+
+    } else if (currentEmailSubscription) {
+      deleteEmailSubscription()
+    }
+    showInSnackbar('Subscription settings saved')
+  }
```
I am not super-happy with this approach: fetching a bunch of API calls and then re-writing the data in them to another place if they exists.  However, I couldn't come up a with a cleaner way of doing this off-hand since on page-load the tabs that aren't active haven't actually loaded this data (so it's not in any state).  Would welcome suggestions on ways to clean this up or make it less... icky.

anathomical committed:
Remove unnecessary times call

anathomical committed:
Fix linting

anathomical committed:
Tweak the way that we copy relevant AdminReportFilterSelections

So that the copy for subscriptions happens on the back-end, not the front-end

anathomical committed:
Update names to be more consistent with other constants

emilia-friedberg commented on services/QuillLMS/app/services/admin_diagnostic_reports/construct_query_payload.rb:
```diff
@@ -6,7 +6,7 @@ class ConstructQueryPayload < ApplicationService
 
     DEFAULT_TIMEFRAME = 'this-school-year'
     DEFAULT_AGGREGATION = 'grade'
-    DEFAULT_DIAGNOSTIC_ID = 1663
```
I appreciate this small cleanup.

emilia-friedberg commented on services/QuillLMS/app/workers/admin_diagnostic_reports/send_monthly_subscriptions_worker.rb:
```diff
@@ -0,0 +1,7 @@
+
+module AdminDiagnosticReports
+  class SendMonthlySubscriptionsWorker < SendSubscriptionsWorker
+    private def subscriptions = EmailSubscription.monthly.premium
```
Love this clean utilization of inheritance.

emilia-friedberg reviewed (approved):
This looks great. I really like the way you handled copying the saved filters to the email filters in the backend -- that pattern feels easily comprehensible and extensible.

anathomical committed:
Update new constant names everywhere they're used

anathomical committed:
Fix linting errors

anathomical committed:
More linting fixes

anathomical committed:
Another place to update constant names

brendanshean commented on services/QuillLMS/app/workers/admin_diagnostic_reports/send_subscriptions_worker.rb:
```diff
@@ -0,0 +1,21 @@
+      end
+    end
+
+    private def subscriptions = (raise NotImplementedError)
```
Oh cool!  I didn't know the syntax for raising exceptions with endless block

anathomical replied:
I think you may be able to get away with not including the parentheses, too, but I feel like it reads a little easier this way.

happythenewsad reviewed (approved):
LGTM

dandrabik commented on services/QuillLMS/app/workers/admin_diagnostic_reports/send_csv_email_worker.rb:
```diff
@@ -4,18 +4,29 @@ module AdminDiagnosticReports
+    private def email_payload
+      @email_payload ||= [
+        user.id,
+        generate_overview_link(user.id, shared_filter_report_name, overview_filter_report_name),
```
The advantage of using `attr_reader` is that you don't need to pass params to methods within the class, since the method will have access to them. This line could be `generate_overview_link` and that method can just use the `attr_reader` methods.

dandrabik commented on services/QuillLMS/app/workers/admin_diagnostic_reports/send_csv_email_worker.rb:
```diff
@@ -4,18 +4,29 @@ module AdminDiagnosticReports
+      ]
     end
 
     private def generate_overview_link(user_id, shared_filter_report_name, overview_filter_report_name)
```
e.g. this should be have no args

```ruby
private def generate_overview_link
  overview_payload = ConstructOverviewQueryPayload.run(user.id, shared_filter_report_name, overview_filter_report_name)

dandrabik reviewed (approved):
Overall, looks good. I left some notes for some code clean up items, but looks good otherwise. Nice work closing this out!

dandrabik commented:
I appreciate the detail in the testing plan. I'm glad to hear this has been thoroughly tested!

anathomical committed:
Update query to sort data after aggregation

For some reason, ORDER BY is reflected in results with a LIMIT clause, but is not reflected in results without LIMIT clauses.

anathomical committed:
Add a comment to explain why we're applying ORDER BY twice

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into adgr-email-subscription

anathomical committed:
Remove unnecessary param passing in service methods

anathomical committed:
Update child workers to not pass params into methods unnecessarily
